### PR TITLE
docs: add design doc

### PR DESCRIPTION
**Pull request by francis36012**

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -1,2 +1,117 @@
...
+
+	google.protobuf.Timestamp created_at = 5;
+	google.protobuf.Timestamp started_at = 6;
+	google.protobuf.Timestamp stopped_at = 7;
+	google.protobuf.Timestamp finished_at = 8;
+}
+
+service JobWorker {
+	// Creates a job using the provided parameters and returns an ID for the created job.
+	rpc CreateJob(CreateJobRequest) returns (Job);
```
Comment (russjones) [inline docs/design.md:71]: This is not needed, just create the job when you start it.

File: docs/design.md [hunk 2]
```diff
@@ -1,2 +1,117 @@
...
+All actions against the job worker service will require authentication. Authentication will be done using mTLS.
+
+## Authorization
+
+There are two types of users recognized by the service:
+
+- Administrator: This type of user can create, start, and stop jobs
+- Normal: This type of user can view the status of jobs
+
+- Q: What's a good way to authorize different user types? Add user type information in client certificate?
```
Comment (russjones) [inline docs/design.md:98]: Yes, putting identity information into the certificate is fine. Where do you plan to store it?

File: docs/design.md [hunk 3]
```diff
@@ -1,2 +1,117 @@
...
+
+The CLI utility can perform all the job actions exposed in the service.  The parameters to the CLI tool will differ with the chosen action but all actions will require an address for the service to connect to. Each call will also require a client certificate to be specified for authentication purposes.
+
+Here are the actions that will be supported by the CLI tool:
+
+- create-job
+- start-job
+- list-jobs
+- get-job-stdout
+- get-job-stderr
```
Comment (russjones) [inline docs/design.md:110]: Can you provide some full examples, like:

```
$ workerctl start-job ls
Created job: 123456789
```

File: docs/design.md [hunk 4]
```diff
@@ -1,2 +1,117 @@
...
+	rpc ListJobs(ListJobsRequest) returns (ListJobsResponse)
+
+	// Get the stdout output of the specified job
+	rpc GetJobStdout(Job) returns (stream bytes);
+}
+```
+
+## Authentication
+
+All actions against the job worker service will require authentication. Authentication will be done using mTLS.
```
Comment (russjones) [inline docs/design.md:89]: Can you add some more details, just a few sentences on how the server will verify the identity of the client and how the client will verify the identity of the server?

Can you also add a section on how you plan to configure TLS?

File: docs/design.md [hunk 5]
```diff
@@ -1,2 +1,117 @@
...
+## Goals
+
+The goals for the the implementation for this library are as follows:
+
+- Provide a core worker library with methods to start, stop, and query the status of a job
+  - For querying job status, the library must support streaming output of job from start of process execution
+- Provide a secure gRPC API, authenticated using mTLS.
+- Provide a simple command line interface for job manipulation
+
+## Core Library Design
```
Comment (russjones) [inline docs/design.md:14]: Can you add a section on how you plan to implement streaming?

File: docs/design.md [hunk 6]
```diff
@@ -1,2 +1,117 @@
 # Job Worker Service / Library
 
+This document details the design of a simple job worker service and library.  The goal of the service is to provide a set of simple interfaces that allow arbitrary linux processes to be started, stopped, and queried.
+
+## Goals
+
+The goals for the the implementation for this library are as follows:
+
+- Provide a core worker library with methods to start, stop, and query the status of a job
```
Comment (russjones) [inline docs/design.md:9]: How will you stop a job?

File: docs/design.md [hunk 7]
```diff
@@ -1,2 +1,117 @@
...
+	// Starts the job with the specified ID.
+	rpc StartJob(Job) returns (Job);
+
+	// Stop the job with the specified ID.
+	rpc StopJob(Job) returns (Job);
+
+	// Get the jobs matching the specified criteria
+	rpc ListJobs(ListJobsRequest) returns (ListJobsResponse)
+
+	// Get the stdout output of the specified job
```
Comment (russjones) [inline docs/design.md:82]: Why only stdout?

Comment (francis36012) [reply docs/design.md:82]: The plan was to include `stderr` but I think `stdout` only should be enough to demonstrate the core functionality.

File: docs/design.md [hunk 8]
```diff
@@ -1,2 +1,117 @@
...
+The goals for the the implementation for this library are as follows:
+
+- Provide a core worker library with methods to start, stop, and query the status of a job
+  - For querying job status, the library must support streaming output of job from start of process execution
+- Provide a secure gRPC API, authenticated using mTLS.
+- Provide a simple command line interface for job manipulation
+
+## Core Library Design
+
+The core library will provide a service for creating, starting, stopping, and querying for the status of jobs. On creation of a job, it's ID is returned to the client. The create call can optionally specify that the created job is started immediately.  Once a job has been created, it can be started (if not started immediately) using the returned ID. A started job can also be stopped before the process completes.
```
Comment (r0mant) [inline docs/design.md:16]: > On creation of a job, it's ID is returned to the client.

What are you planning to use as a job ID?

File: docs/design.md [hunk 9]
```diff
@@ -1,2 +1,117 @@
...
+syntax = "proto3";
+
+message JobProcessParameters {
+	string process = 1;
+	repeated string process_parameters = 2;
+}
+
+message CreateJobRequest {
+	JobProcessParameters process_parameters = 1;
+	bool start_on_create = 2;
```
Comment (r0mant) [inline docs/design.md:32]: Let's just make this true by default and get rid of this field.

File: docs/design.md (hunk 7 above)
Comment (r0mant) [reply docs/design.md:82]: Streaming both stdout and stderr is a part of the requirements. Could you please update the API to support it?

**Review by russjones — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

**Pull request merged**